[댓글 기능] 에단 미션 제출합니다.

Laterality opened the pull request.

gch01410 committed:
[게시글 생성/조회/수정/삭제] 던 미션 제출합니다. (#40)

* feat: Set gradle and create templates folder

* feat: Map article-edit.html on writing url

* feat: Add the function of writing articles

* refactor: Change the setting from the relative path to the absolute path

* feat: Add the function of checking the list of articles

* feat: Map each article on permalink

* refactor: Eliminate duplication of html files

* refactor: Modify path and optimize the name of variable

* feat: Add function of editing articles

* feat: Add function of deleting articles

* refactor: Remove duplication of html

* refactor: Merge MainController to ArticleController

* refactor: Create articleDTO

* refactor: Create exception handler of not finding to article

* refactor: Separate the functions into service layers from controller layers

* refactor: Remove MainControllerTest

* refactor: Fix the name of method to apply method naming convention

* refactor: Remove the setter method of DTO

* refactor: Create assembler of Article to convert to dto and entity

* refactor: Move latest id of article to ArticleRepository

* refactor: Customize the exception about Article

* refactor: Replace redirect string with RedirectView

* refactor: Add ResponseStatus annotation to exception

* refactor: Rename package of exception

gch01410 committed:
[게시판 회원 기능] 던 미션 제출합니다. (#81)

* delete: Delete ArticleRepository memory version

* refactor: Fix testcodes to accomodate repository change

* feat: Show to login view and sign up view

* feat: Insert user data to DB

* feat: Validate sign up data

* feat: Add redirection when sign up fails

* refactor: Add duplicated email exception

* feat: Show user list

* feat: Add user login function

* feat: Add user login error message

* feat: Customize the login exception

* feat: Add user logout function

* feat: Limit access when logged in

* feat: Show user information in my page

* feat: Add the function to modify user information

* feat: Add the function to remove user information

* refactor: Validate sign up information in front end

* refactor: Add final keyword to parameter and field

* refactor: Create login service layer and Separate domain and service package

* feat: Validate in entity and replace id type with Long

* test: Add login service test and user service test

* refactor: Replace method to check null with Objects.requireNonNull

* refactor: Replace method to check null with Objects.requireNonNull

* refactor: Remove if statement

* test: Remove duplicated block

* feat: Add controller advice to handle exception

* feat: Add controller advice to handle exception

* feat: Add interceptor class

* test: Add test code

* style: Create constant of user session key

* refactor: Separate packages

* feat: Validate sign up information in backend

Commit:
docs: Update README.md

Commit:
feat: Map between user and article

Commit:
feat: Implement permission about article edit

Commit:
feat: Add comment domain

Commit:
feat: Add comment service

Commit:
feat: Add comment controller

vsh123 committed:
Revert "[게시판 회원 기능] 에단 미션 제출합니다 (#48)"

This reverts commit d6c8a1a88574616ab8c354c54e891ac9491fbfc0.

vsh123 committed:
Revert "[게시글 생성/조회] 에단 미션 제출합니다. (#2)"

This reverts commit 095864c41b1d37f03f079597b55ed8ddfeec2334.

Laterality committed:
Update web config

- Added `/bin` directory to `.gitignore`.

Laterality committed:
Update properties

Laterality committed:
Refactor controller test

woowahan-pjs reviewed (changes requested):
오프라인 코드 리뷰로 대신하였습니다. 👍

Laterality committed:
Refactor test

Laterality committed:
Apply feedback

- Changed access modifier.

Laterality committed:
Apply feedback: refactor service

Laterality committed:
Change article contents to lob type

Laterality committed:
Refactor: code format

Laterality committed:
Refactor: Change to return `String` instead of `ModelAndView`

Laterality committed:
Refactor: remote unnecessary parenthesis

Laterality committed:
Refactor: add method matching dto's id

Laterality committed:
Refactor domain

- Added `@CreatedDate` and `@LastModifiedDate`.
- Added cascade property.

Laterality committed:
Refactor

- Resolved failing tests.
- Remoted lines no more needed.

Laterality committed:
Refactor: format code

Laterality committed:
Resolve failed tests

Laterality committed:
Refactor: remove unnecessary methods

Laterality committed:
Update: enable jpa auditing for creating created date and last modified date automatically

Laterality committed:
Refactor to use pre-defined constant

Laterality commented:
다른 리뷰 내용들과 오프라인 리뷰 때 생각나는 부분들을 돌아보면서 수정해봤습니다.

컨트롤러 메서드의 반환 타입을  `ModelAndView`에서 `String`으로 바꿔서 코드를 간소화했고, 
`CommentControllerTest`의 메서드를 리팩터링해보고 있습니다. 

아직 테스트 코드의 리팩터링이 미흡하다고 생각하지만 프로덕션 코드에 대한 리뷰를 먼저 받아보고 싶어 리뷰 요청드립니다!

Laterality committed:
Refactor login controller tests

- Removed unnecessary annotation.
- Segregated code requesting login.

Laterality committed:
Refactor: Changed package structure

Laterality committed:
Merge branch 'lateralitys' of https://github.com/Laterality/jwp-blog into lateralitys

Laterality committed:
Reafactor: rename common block files

Laterality committed:
Refactor template

- Segregated header.
- Segregated footer.
- Changed productions property to use mysql.
- Changed user entity's table name to avoid using mysql's reserved word.

Laterality committed:
Refactor template

- Changed article delete button's style.
- Chnaged comment edit button style.

Laterality committed:
Add interceptor for putting user into model

Laterality committed:
Refactor: remove unnecessary code

Laterality committed:
Update template: user deletion button style and location

Laterality committed:
Refactor article service

- Changed interceptor path pattern mapping.
- Resolved test failed as of the changes.

Laterality committed:
Refactor services to return domain objects

Laterality committed:
Refactor: format code

woowahan-pjs reviewed (changes requested):
피드백을 잘 반영해 주셨습니다. 👍 
간단한 생각거리를 남겼으니 코드에 반영해 주셔도 좋습니다.
진행 과정에서 궁금한 점은 댓글이나 Slack을 통해 물어보세요.

Laterality committed:
Update build.gradle

- Changed Mysql jdbc to 'runtimeOnly' from 'implementation'.

Laterality committed:
Refactor: restructure package organization

- Removed unnecessary code.
- Renamed `ResponseDto` to `Dto`.
- Changed method returning collection to return unmodifiable list.

Laterality committed:
Update article service test to use mockito

Laterality committed:
Refactor

- Added exceptions for deleting/editing article and comment.
- Updated comment tests to use mockito.

Laterality commented:
리뷰 감사합니다!

패키지 구조를 조정하고 컨트롤러의 로직을 서비스로 이동했습니다.

mockito를 활용해서 `ArticleServiceTest`와 `CommentServiceTest`에 먼저 적용해봤는데, Mock 객체를 초기화하는 코드가 상당히 긴 것 같아서 잘 작성한 건지 모르겠네요.

예외처리를 추가하면서 궁금한 점이 하나 생겼습니다. `/articles/{articleId}`를 처리하는 컨트롤러에서 예외가 발생한 경우 `ArticleConrollerAdvice`에서 해당 예외에 대한 핸들러를 찾아 응답하고 있습니다. 하지만 핸들러는 컨트롤러에서 인자로 받을 값을 알 수가 없는 것으로 보입니다. 이런 경우 핸들러에서 `"redirect:/articles/" + articleId`와 같은 형태로 리다이렉트할 수 있는 방법이 있을까요?

Laterality committed:
Refactor service tests to use default objects

Laterality committed:
Refactor: organized application properties

woowahan-pjs reviewed (changes requested):
간단한 생각거리를 남겼으니 코드에 반영해 주셔도 좋습니다.
진행 과정에서 궁금한 점은 댓글이나 Slack을 통해 물어보세요.

Laterality committed:
Apply feedback

- Updated H2 dependency to runtime dependency.
- Removed 'XXXAssembler'.

Laterality committed:
Refactor service test with mockito

Laterality committed:
Refactor: remove unnecessary spying

Laterality commented:
엔티티 객체들의 ID 처리를 어떻게 할 것인지에 관해 낮에 말씀하셨던 것들을 포함해서 다양한 방법들을 고민해봤습니다. 각자 나름의 이유는 있지만 **이거다!** 싶은 대안이 떠오르지 않아서 `spy()`를 적용해봤습니다. 객체를 부분적으로 mocking할 수 있다는 점에서 지금의 상황에 필요하다고 생각했는데요.

주로 다음 방식들을 고민해봤습니다.

* 엔티티 객체를 부분적으로 Mocking하는 *어떤(마법?)* 방법
* 엔티티 객체에 `setId()` 추가
* 엔티티 객체를 상속받는 테스트용 엔티티 객체를 생성하고 `setId()`를 추가한 뒤,
  ```java
  User defaultUser = mock(TestUser.class);
  ```
  와 같이 사용

이 케이스가 `spy()`를 사용하기 적절한 상황인지는 계속 의문이 생기지만  생각해본  방법들 중에서는 
 추가로 클래스를 작성할 필요가 없고 의도했던 테스트 코드와 비슷한 흐름으로 간다는 점에서 최선책이라고 판단했습니다.

woowahan-pjs reviewed (approved):
수많은 피드백을 반영하고자 대단히 고생 많으셨습니다. 💯  
좋은 개발자가 되기 위한 소중한 밑거름이 되었길 바랍니다.
이만 Merge 하겠습니다.
